qemu/tests/qsd_backup_push.py

```diff
@@ -0,0 +1,152 @@
+            image_params = params.object_params(tag)
+            image_chain = image_params.objects("image_backup_chain")
+            self.source_images.append("fmt_%s" % tag)
+            self.full_backups.append("drive_%s" % image_chain[0])
+            self.inc_backups.append("drive_%s" % image_chain[1])
+            self.bitmaps.append("bitmap_%s" % tag)
+            inc_img_tag = image_chain[-1]
+            inc_img_params = params.object_params(inc_img_tag)
+
+            # rebase 'inc' image onto 'base' image, so inc's backing is base
+            inc_img_params['image_chain'] = image_params['image_backup_chain']
+            inc_img = self.source_disk_define_by_params(
+                inc_img_params, inc_img_tag)
+            target_func = partial(inc_img.rebase, params=inc_img_params)
+            self.rebase_targets.append(target_func)
+
+    def get_qsd_demon(self):
+        qsd_name = self.params["qsd_namespaces"]
+        qsd_ins = QsdDaemonDev(qsd_name, self.params)
+        return qsd_ins
+
+    def start_qsd(self):
+        self.qsd = self.get_qsd_demon()
+        self.qsd.start_daemon()
+
+    @error_context.context_aware
+    def add_target_data_disks(self):
+        """Hot add target disk to VM with qsd monitor"""
+        error_context.context("Create target disk")
+        for tag in self.params.objects("source_images"):
+            image_params = self.params.object_params(tag)
+            for img in image_params.objects("image_backup_chain"):
+                disk = self.target_disk_define_by_params(self.params, img)
+                disk.hotplug(self.qsd)
+                self.trash.append(disk)
+
+    def do_full_backup(self):
+        extra_options = {"sync": "full", "auto_disable_bitmap": False}
+        backup_utils.blockdev_batch_backup(
+            self.qsd,
+            self.source_images,
+            self.full_backups,
+            self.bitmaps,
+            **extra_options)
+
+    def generate_inc_files(self):
+        for tag in self.params.objects("source_images"):
+            self.generate_data_file(tag)
+
+    def do_increamental_backup(self):
+        extra_options = {'sync': 'incremental', 'auto_disable_bitmap': False}
+        backup_utils.blockdev_batch_backup(
+            self.qsd,
+            self.source_images,
+            self.inc_backups,
+            self.bitmaps,
+            **extra_options)
+
+    def rebase_target_disk(self):
+        self.qsd.stop_daemon()
+        return utils_misc.parallel(self.rebase_targets)
+
+    def prepare_clone_vm(self):
+        self.main_vm.destroy()
+        images = self.params["images"]
+        clone_params = self.main_vm.params.copy()
+        for tag in self.params.objects("source_images"):
+            img_params = self.params.object_params(tag)
+            image_chain = img_params.objects('image_backup_chain')
+            images = images.replace(tag, image_chain[-1])
+            self.params["qsd_images_qsd1"] = image_chain[-1]
```

Comment:
outside for loop: 
qsd_images = []
inside for loop:
qsd_images.append(image_chain[-1])
outside for loop:
self.params["qsd_images_qsd1"] = ' '.join(qsd_images)

We handle the source images as a list, so the inc images can also be a list here